### PR TITLE
Simplify prodElectrons and prodMuons

### PR DESCRIPTION
**Pull request by hatakeyamak**
+ modified so that prodMuons and prodElectrons run once, and generate different lists for storing muons/electrons and for top projection

### CODE REVIEW

**Comment by hatakeyamak** [comment]
You see inputTag changes in SkimsAUX/python/prodJets_cfi.py.

In prodElectrons.cc
    if( isVetoID && miniIso < maxEleMiniIso_ && ele->pt() > minElePtForElectron2Clean_ 
&& fabs(ele->eta()) < maxEleEta_) ele2Clean->push_back(*ele);
This eta requirement was added (|eta|<2.5). This eta requirement is not applied to electrons stored in ntuples. Is it intended? Right now, pt candidate's eta is used for the selection. Another possibility is to cut on the associated supercluster's eta (but I didn't touch this part for now).

In prodMuons,cc
I didn't see any need to adjust the muon selection to be used for top projections. |eta|<2.4 seems to be used by default.

By the way, for top projection lepton selection, pt>10 GeV seems to be used, while pt>5 GeV for storing leptons and we plan to use leptons down to 5 GeV for veto. Just checking this 10 GeV cut for top projection leptons is intended.

**Comment by pastika** [comment]
We need to clean what we select, not what we veto, so I think 10 GeV makes sense.  This way we do not clean out low pT muons which can have a much higher misID rate.  The lowest pT muon we select is 24 GeV usually, so we should be ok.

File: SkimsAUX/workdir/treeMaker_stopRA2.py [hunk 1]
```diff
@@ -492,13 +486,13 @@
    process.stopTreeMaker.varsBool.append(cms.InputTag("HBHENoiseFilterResultProducer", "HBHEIsoNoiseFilterResult"))
    process.stopTreeMaker.varsBoolNamesInTree.append("HBHENoiseFilterResultProducer:HBHEIsoNoiseFilterResult|HBHEIsoNoiseFilter")
 
+process.stopTreeMaker.varsInt.append(cms.InputTag("prodMuons", "nMuons_cut"))
```
Comment (pastika) [inline SkimsAUX/workdir/treeMaker_stopRA2.py]: I thought we were removing this variable?

File: SkimsAUX/workdir/treeMaker_stopRA2.py [hunk 2]
```diff
@@ -556,15 +550,15 @@
 process.stopTreeMaker.vectorTLorentzVector.append(cms.InputTag("prodSecondaryVertex", "svLVec"))
 
 
+process.stopTreeMaker.varsInt.append(cms.InputTag("prodElectrons", "nElectronsCut"))
```
Comment (pastika) [inline SkimsAUX/workdir/treeMaker_stopRA2.py]: We should remove this variable too?

File: SkimsAUX/plugins/prodGenInfo.cc [hunk 3]
```diff
@@ -103,6 +107,12 @@ prodGenInfo::prodGenInfo(const edm::ParameterSet & iConfig) {
   produces< std::vector< TLorentzVector > >("selGenParticle"); 
   produces< std::vector< int > >("selPDGid");
 
+  // 2017 94x numbers (just to make this code work)
+  genEleEAValues={0.1440, 0.1562, 0.1032, 0.0859, 0.1116, 0.1321, 0.1654};
+  genEleEAEtaValues={1.0,    1.479,    2.0,    2.2,    2.3,    2.4};
+  genMuEAValues={0.0735, 0.0619, 0.0465, 0.0433, 0.0577};
+  genMuEAEtaValues={0.8,    1.3,    2.0,    2.2};
```
Comment (zhenbinwu) [inline SkimsAUX/plugins/prodGenInfo.cc]: These values are hard coded for 2017. Not sure whether this remain the same for 2016 & 2018. I am OK with it for now. Next round we should try to update the code for capability of 3 years' configs.

File: SkimsAUX/workdir/treeMaker_stopRA2.py (hunk 1 above)
Comment (hatakeyamak) [reply SkimsAUX/workdir/treeMaker_stopRA2.py]: ok. will do.

File: SkimsAUX/workdir/treeMaker_stopRA2.py (hunk 2 above)
Comment (hatakeyamak) [reply SkimsAUX/workdir/treeMaker_stopRA2.py]: ok. will do.

File: SkimsAUX/plugins/common.cc [hunk 4]
```diff
@@ -117,4 +135,13 @@ namespace commonFunctions
       return activity;
   }
 
+  double getEta(const reco::Candidate& obj) {
+    try{
+      const pat::Electron& electron = dynamic_cast<const pat::Electron&>(obj);
```
[the remark this answers is not in the document]
Comment (hatakeyamak) [reply SkimsAUX/plugins/common.cc]: this line should be used only for ele. if muon is an input, the second block should take its effect.

File: SkimsAUX/plugins/prodGenInfo.cc (hunk 3 above)
Comment (scarletnorberg) [reply SkimsAUX/plugins/prodGenInfo.cc]: Yes. I think there should just be an if statement for all three years or an array for every year? But this does not need to be done right this second.

Comment (hatakeyamak) [reply SkimsAUX/plugins/prodGenInfo.cc]: I did that for prodEle and prodMu. For here, for gen, I am not sure what is the most optimal, and I don't see why this should be year-dependent/time-dependent. Let's just keep it for now.
I don't think we actually use activity varaibles for gen in a critical way anyway.

**Comment by hatakeyamak** [comment]
Implemented Joe's suggestions to remove "cut" variables. A short test run looked successful.
/uscms_data/d2/hatake/HadronicStop/CMSSW_9_4_10_LPC/src/StopTupleMaker/SkimsAUX/workdir/stopFlatNtuples_100events.root
is the output of the test.
The full event test job is running, and output will be:
/uscms_data/d2/hatake/HadronicStop/CMSSW_9_4_10_LPC/src/StopTupleMaker/SkimsAUX/workdir/stopFlatNtuples.root
I think the job takes about 12 hours.

**Comment by zhenbinwu** [comment]
12 hours? For the same MINIAOD file we have been using?

**Comment by scarletnorberg** [comment]
This is expected. Matt does not have deepAK8 and does not do the jet reclustering yet. Those are the time synchs. DeepAK8 there is nothing to be done because MXNet is slow. We have done what we think we can do with the jet code as of now. We know that CversusL is also slow which Matt does not need because it is used as comparison between the 49 BDT resolved and Joes resolved tagger. We will remove this variable in the future. It was also suggested to maybe go back to doing things by hand for photon ID flag. The photon ID code is quiet slow in CMSSW. This is not the largest time synch though.

**Comment by zhenbinwu** [comment]
Ok, merging now

**Comment by hatakeyamak** [comment]
Zhenbin did you check ntuples with 100 events which I sent you earlier? It will provide a very useful check already.

**Pull request merged**